File: 2025-and-before/0112-path-sum/0112-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */

const helper = function(root, obj){
    
    if(!root)
        return false;
    
    let ans;
    obj.targetSum-=root.val;
    
    if(!root.left && !root.right)
        ans =  obj.targetSum == 0? true : false;
    
    
    else    
        ans = helper(root.left,obj) || helper(root.right,obj);
    
    obj.targetSum += root.val;
    
    return ans;
}
var hasPathSum = function(root, targetSum) {
    let obj = {targetSum : targetSum};
    
    return helper(root,obj);
};

File: 2025-and-before/0112-path-sum/0112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution{
	
	private: 
	bool isLeaf(TreeNode *root){
		if(!root)
		return false;
		

		if(!root->left and !root->right)
		return true;

		return false;
}



	public :

	bool helper(TreeNode *root, int&target, int&currSum){
		
		if(!root)
		return false;

		currSum+=root->val;
		
		if(isLeaf(root) and currSum==target)
		return true;
		
		bool ans = false | helper(root->left,target,currSum) | helper(root->right,target,currSum);
		currSum-=root->val;
		return ans;	
}

	bool hasPathSum(TreeNode *root,int targetSum){

	int currSum = 0;
	return helper(root,targetSum,currSum);
}

};
